music21/osmd/osmd.py

```diff
@@ -0,0 +1,222 @@
+See OSMD project page here: https://github.com/opensheetmusicdisplay/opensheetmusicdisplay
+'''
+import unittest
+import time
+import random
+import json
+import os
+import importlib
+import tempfile
+import urllib.request
+import webbrowser
+
+from music21.converter.subConverters import SubConverter
+from music21.instrument import Piano
+from music21.common import getSourceFilePath
+from music21.common import runningUnderIPython
+from music21 import exceptions21
+
+
+class OpenSheetMusicDisplayException(exceptions21.Music21Exception):
+    pass
+
+
+try:
+    loader = importlib.util.find_spec('IPython.core.display')
+except ImportError:
+    loader = None
+hasInstalledIPython = loader is not None
+del importlib
+
+
+def getExtendedModules():
+    # pylint: disable=import-outside-toplevel
+    if hasInstalledIPython:
+        from IPython.core.display import display, HTML, Javascript
+    else:
+        def display():
+            raise OpenSheetMusicDisplayException(
+                'OpenSheetMusicDisplay requires IPython to be installed')
+
+        HTML = Javascript = display
+    return display, HTML, Javascript
+
+
+class ConverterOpenSheetMusicDisplay(SubConverter):
+    '''
+    Converter for displaying sheet music inline
+    in an IPython notebook.
+    Uses: https://opensheetmusicdisplay.org/ for rendering
+    '''
+    registerFormats = ('osmd',)
+    registerShowFormats = ('osmd',)
+    # when updating the script tag osmd will only reload in a notebook if you: Kernel -> restart and
+    # clear output, then save the notebook and refresh the page.
+    # Otherwise the script will stay on the page and not reload.
+    script_url = ("https://github.com/opensheetmusicdisplay"
+                  + "/opensheetmusicdisplay/releases/download/0.9.2/opensheetmusicdisplay.min.js")
+    osmd_file = os.path.join(getSourceFilePath(), 'osmd', 'opensheetmusicdisplay.0.9.2.min.js')
+
+    def __init__(self):
+        super().__init__()
+        self.display, self.HTML, self.Javascript = getExtendedModules()
+
+    def show(self, obj, fmt,
+             fixPartName=True, offline=False, divId=None,
+             **keywords):
+        r'''
+        Displays the score object in a notebook using the OpenSheetMusicDisplay.js library.
+
+        >>> import music21
+        >>> s = music21.converter.parse("tinyNotation: 3/4 E4 r f# g=lastG trip{b-8 a g} c4~ c")
+        >>> fig_id1 = s.show('osmd')
+
+        To update a previously displayed score use the returned divId from before:
+
+        >>> fig_id2 = s.show('osmd', divId=fig_id1)
+        >>> assert fig_id1 == fig_id2
+        '''
+        in_ipython = runningUnderIPython()
+        score = obj
+        if fixPartName:
+            self.addDefaultPartName(score)
+
+        if divId is None:
+            # create unique reference to output div in case we wish to update it
+            divId = self.getUniqueDivId()
+
+
+        xml = open(score.write('musicxml')).read()
```

Comment:
is this file ever closed? perhaps better to use a `with open(...) as f:` context manager?

Reply:
Actually, if you just want to get some xml into memory, you can just use the `GeneralObjectExporter` in the musicxml export class; this is a bit safer than depending on the user having permissions to write to a temp dir

```
gex = music21.musicxml.m21ToXml.GeneralObjectExporter(score)
bytesOut = gex.parse()
xml = bytesOut.decode('utf-8')
```